dask/dataframe/core.py

```diff
@@ -4132,6 +4181,38 @@ def applymap(self, func, meta="__no_default__"):
     def round(self, decimals=0):
         return elemwise(M.round, self, decimals)
 
+    @derived_from(pd.DataFrame)
+    def mode(self, axis=0, numeric_only=False, dropna=True, split_every=False):
+        axis = self._validate_axis(axis)
+        if axis == 1:
+            return super(DataFrame, self).mode(
+                axis=axis, dropna=dropna, split_every=split_every
+            )
+        elif axis == 0:
+            mode_series_list = []
+            for col in self.columns:
+                mode_series = super(Series, self[col]).mode(
+                    dropna=dropna, split_every=split_every
+                )
+                mode_series.name = col
+                mode_series_list.append(mode_series)
+
+            name = "concat-" + tokenize(*mode_series_list)
+
+            concat_axis1 = partial(methods.concat, axis=1)
```

Comment:
We try to avoid placing partials and other dynamicly generated functions into the task graph.  This aids future debugging and reduces scheduling overhead.  If you need to add a keyword, then consider using the `apply` function.  

However, I also wonder if there is a way to do what you're trying to do here with `map_partitions` instead.  If 
so, it would be good to do so.  More future devs/reviewers will be able to understand `map_partitions` than custom graphs (which are used only when we need to do something complicated).

Reply:
I replaced partial with apply.  

I don't think map_partitions is appropriate here though I'm happy to try additional suggestions.

**Why I don't believe map_partitions is appropriate here**
In this implementation, mode is computed on each series of the dataframe.  As part of computing mode on a series, all of the Series partitions are collapsed into a single partition via `value_counts` and `sum`.  This last bit of the mode method is trying to concatenate the result of having run mode on each of the series.  Each of the series are often of differing lengths, and I am trying to concatenate them along axis=1 to construct the final dataframe to be returned with nans filling in the extra space in the shorter series (as pandas does).

The `concat_unindexed_dataframes` in `dask/dataframe/multi.py` (see below) is very close to what I want to do except it calls `concat_and_check` which checks that the series are the same length, which they are not.  I therefore copied `concat_and_check` and modified it slightly to do what I want in this bit of code.

**Part of `dask/dataframe/multi.py` for reference**
```python3
def concat_and_check(dfs):
    if len(set(map(len, dfs))) != 1:
        raise ValueError("Concatenated DataFrames of different lengths")
    return methods.concat(dfs, axis=1)


def concat_unindexed_dataframes(dfs):
    name = "concat-" + tokenize(*dfs)

    dsk = {
        (name, i): (concat_and_check, [(df._name, i) for df in dfs])
        for i in range(dfs[0].npartitions)
    }

    meta = methods.concat([df._meta for df in dfs], axis=1)

    graph = HighLevelGraph.from_collections(name, dsk, dependencies=dfs)
    return new_dd_object(graph, name, meta, dfs[0].divisions)
```